cpp/Tuparg.hpp

```diff
@@ -0,0 +1,94 @@
+#pragma once
+
+#include "Types.hpp"
+#include "Utility.hpp"
+#include <cassert>
+
+namespace hsd {
+    struct xtuple_iterator {
+        const usize* layout;
+        void* addr;
+
+        template <typename X>
+        X& getnext() {
+            assert(addr);
+
+            X* p = reinterpret_cast<X*>(addr);
+            addr = reinterpret_cast<void*>(
+                reinterpret_cast<uptr>(addr)
+                + *layout
+            );
+            if (* ++layout == 0)
+                addr = nullptr;
+            return *p;
+        }
+
+        bool operator==(const xtuple_iterator& o) const {
+            return addr == o.addr;
+        }
+
+        bool operator!=(const xtuple_iterator& o) const {
+            return addr != o.addr;
+        }
+    };
+
+    template <typename Tfirst, typename... Trest>
+    struct xtuple {
+        Tfirst _first;
+        xtuple<Trest...> _rest;
+
+        xtuple(Tfirst f, Trest... r)
+            : _first(move(f)), _rest(move(r)...)
+        {}
+
+        consteval static usize get_offset() {
+            return offsetof(xtuple, _rest);
```

Comment:
Can we implement this?

Reply:
no, but the closest thing we can do is `#define offsetof __builtin_offsetof`